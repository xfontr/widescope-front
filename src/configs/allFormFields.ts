import {
  FormComponents,
  FormFields,
} from "../components/RenderForm/RenderFormTypes";

export interface FormValue {
  type?: React.HTMLInputTypeAttribute;
  label: string;
  name: FormFields;
  belongsTo: FormComponents[];
  placeholder?: string;
  customClass?: string;
  customGroupClass?: string;
  renderAs?: "input" | "textarea";
}

export const allFormFields: Record<FormFields, FormValue> = {
  name: {
    type: "text",
    label: "Name",
    name: "name",
    belongsTo: ["logIn", "signUp"],
    placeholder: "John Doe",
    customClass: "form__input--user",
  },

  password: {
    type: "password",
    label: "Password",
    name: "password",
    belongsTo: ["logIn", "signUp"],
    customClass: "form__input--password",
  },

  repeatPassword: {
    type: "password",
    label: "Repeat password",
    name: "repeatPassword",
    belongsTo: ["signUp"],
  },

  email: {
    type: "email",
    label: "Email address",
    name: "email",
    belongsTo: ["signUp"],
  },

  projectName: {
    type: "text",
    label: "Name",
    name: "name",
    placeholder: "Music app",
    belongsTo: ["createProject", "updateProject"],
  },

  repository: {
    type: "text",
    label: "Repository URL",
    name: "repository",
    placeholder: "http://www.github.com/music-app",
    belongsTo: ["createProject", "updateProject"],
  },

  logo: {
    type: "file",
    label: "Project logo",
    name: "logo",
    belongsTo: ["createProject"],
    customGroupClass: "area",
  },

  logoUpdate: {
    type: "file",
    label: "Project logo",
    name: "logoUpdate",
    belongsTo: ["updateProject"],
    customGroupClass: "area",
  },

  technologyFront: {
    type: "text",
    label: "Frontend main library or framework",
    name: "technologyFront",
    placeholder: "React",
    belongsTo: ["createProject", "updateProject"],
  },

  technologyBack: {
    type: "text",
    label: "Backend main library or framework",
    name: "technologyBack",
    placeholder: "React",
    belongsTo: ["createProject", "updateProject"],
  },

  description: {
    label: "Description",
    name: "description",
    placeholder: "Music app is about sharing music between users",
    belongsTo: ["createProject", "updateProject"],
    renderAs: "textarea",
    customGroupClass: "area",
  },
};
