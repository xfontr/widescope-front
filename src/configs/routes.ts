const routes = {
  root: "/",
  home: "/home",
  signUp: "/sign-up",
  logIn: "/log-in",
  explore: "/explore",
  projectDetails: "/project/:projectId",
};

export default routes;
