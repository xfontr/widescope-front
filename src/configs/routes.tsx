import React from "react";
import { Navigate } from "react-router-dom";
import LandingPage from "../pages/LandingPage/LandingPage";
const ExplorePage = React.lazy(
  () => import("../pages/ExplorePage/ExplorePage")
);
const LogInPage = React.lazy(() => import("../pages/LogInPage/LogInPage"));
const ManageProjectPage = React.lazy(
  () => import("../pages/ManageProjectPage/ManageProjectPage")
);
const NotFoundPage = React.lazy(
  () => import("../pages/NotFoundPage/NotFoundPage")
);
const ProjectDetailsPage = React.lazy(
  () => import("../pages/ProjectDetailsPage/ProjectDetailsPage")
);
const SignUpPage = React.lazy(() => import("../pages/SignUpPage/SignUpPage"));
const UserProjectsPage = React.lazy(
  () => import("../pages/UserProjectsPage/UserProjectsPage")
);

export interface NavRoute {
  path: string;
  display: "never" | "always" | "notLogged" | "logged";
  name: string;
  skip?: boolean;
  action?: (action: () => void) => void;
  render?: () => JSX.Element;
}

export const navRoutes: Record<string, NavRoute> = {
  notFound: {
    path: "*",
    display: "never",
    name: "Not found",
    render: () => <NotFoundPage />,
  },

  root: {
    path: "/",
    display: "never",
    name: "Home",
    render: () => <Navigate to={navRoutes.home.path} />,
  },

  home: {
    path: "/home",
    display: "always",
    name: "Home",
    render: () => <LandingPage />,
  },

  explore: {
    path: "/explore",
    display: "always",
    name: "Explore",
    render: () => <ExplorePage />,
  },

  signUp: {
    path: "/sign-up",
    display: "notLogged",
    name: "Sign up",
    render: () => <SignUpPage />,
  },

  logIn: {
    path: "/log-in",
    display: "notLogged",
    name: "Log in",
    render: () => <LogInPage />,
  },

  logOut: {
    path: "",
    display: "logged",
    name: "Log out",
    action: (action) => action(),
  },

  personalProjects: {
    path: "/personal-projects",
    display: "logged",
    name: "Your projects",
    render: () => <UserProjectsPage />,
  },

  createProject: {
    path: "/project/new",
    display: "logged",
    skip: true,
    name: "Post a project",
    render: () => <ManageProjectPage isCreate={true} />,
  },

  updateProject: {
    path: "/projects/update/:projectId",
    display: "never",
    name: "Update project",
    render: () => <ManageProjectPage isCreate={false} />,
  },

  projectDetails: {
    path: "/project/:projectId",
    display: "never",
    name: "Project details",
    render: () => <ProjectDetailsPage />,
  },
};
