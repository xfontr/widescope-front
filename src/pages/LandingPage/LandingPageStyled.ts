import styled from "styled-components";

export const HeroSectionStyled = styled.section`
  background-image: linear-gradient(
      to bottom,
      rgba(255, 255, 255, 0.7),
      rgba(255, 255, 255, 0.7)
    ),
    url("/img/background.jpeg");

  background-attachment: fixed;
  background-position: top left;
  margin-top: -12.4vh;
  margin-bottom: -10vh;
  padding: 14vh 0 20vh 1.4rem;

  .container {
    max-width: ${(props) => props.theme.breakpoints.verySmall};
    margin: 0 auto;

    @media (min-width: ${(props) => props.theme.breakpoints.verySmall}) {
      max-width: ${(props) => props.theme.breakpoints.big};
    }

    @media (min-width: ${(props) => props.theme.breakpoints.big}) {
      max-width: ${(props) => props.theme.breakpoints.large};
    }
  }

  &:first-child a {
    margin-right: ${({ theme }) => theme.spacing.gapSmall};
  }

  .page__title {
    &--landing {
      font-size: clamp(3.5rem, 5.5vw, 7rem);
      line-height: 4rem;
      margin-bottom: ${({ theme }) => theme.spacing.gapBig};

      @media (min-width: ${({ theme }) => theme.breakpoints.large}) {
        max-width: 70%;
      }
    }

    &-subheading--landing {
      font-size: 1.3rem;
      margin-bottom: ${({ theme }) => theme.spacing.gapBig};

      @media (min-width: ${({ theme }) => theme.breakpoints.large}) {
        max-width: 40%;
      }
    }
  }
`;

export const CTASectionStyled = styled.section`
  background-color: ${({ theme }) => theme.colors.secondary};
  margin-top: 5rem;
  padding-top: 1rem;
  padding-bottom: 5rem;
  color: ${({ theme }) => theme.colors.primary};

  .container {
    padding: 1.4rem;
    max-width: ${(props) => props.theme.breakpoints.verySmall};
    margin: 0 auto;

    @media (min-width: ${(props) => props.theme.breakpoints.verySmall}) {
      max-width: ${(props) => props.theme.breakpoints.big};
    }

    @media (min-width: ${(props) => props.theme.breakpoints.big}) {
      max-width: ${(props) => props.theme.breakpoints.large};
    }
  }

  h2 {
    margin-bottom: 0;
  }

  .landing-page__log-in {
    padding-top: 1rem;
    .page__title {
      padding-bottom: 1rem;
    }
    @media (min-width: ${({ theme }) => theme.breakpoints.large}) {
      .page__title {
        max-width: 100%;
      }
    }
  }

  .landing-page__create {
    padding-top: ${({ theme }) => theme.spacing.gapBig};

    .page__title {
      margin-bottom: ${({ theme }) => theme.spacing.gapSmall};
    }

    @media (min-width: ${({ theme }) => theme.breakpoints.large}) {
      text-align: center;
      .page__title {
        max-width: 100%;
      }
    }
  }
`;
