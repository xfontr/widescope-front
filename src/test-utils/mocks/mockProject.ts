import { IProject } from "../../types/project";

const mockProject: IProject = {
  id: "630e31ecb968115ba6a3e29f",
  name: "Project123",
  description: "The description",
  technologies: ["React", "Express"],
  repository: "www.google.com",
  author: "username",
  authorId: "id",
  logo: "ProjectLogo.png",
  logoBackup: "BackupProjectLogo.png",
  creationDate: "1662627402346",
};

export default mockProject;
