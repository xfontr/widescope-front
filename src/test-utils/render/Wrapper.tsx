import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import { store } from "../../app/store";
import GlobalStyles from "../../styles/GlobalStyles";
import styledMainTheme from "../../styles/styledMainTheme";

interface WrapperProps {
  children: JSX.Element | JSX.Element[];
}

const Wrapper = ({ children }: WrapperProps): JSX.Element => {
  return (
    <ThemeProvider theme={styledMainTheme}>
      <Provider store={store}>
        <BrowserRouter>
          <GlobalStyles />
          {children}
        </BrowserRouter>
      </Provider>
    </ThemeProvider>
  );
};

export default Wrapper;
