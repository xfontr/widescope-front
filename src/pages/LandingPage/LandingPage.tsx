import { lazy, Suspense, memo, useEffect } from "react";
import { useAppSelector } from "../../app/hooks";
import Button from "../../components/Button/Button";
import useProjects from "../../hooks/useProjects/useProjects";
import { CTASectionStyled, HeroSectionStyled } from "./LandingPageStyled";
import { useNavigate } from "react-router-dom";
import { navRoutes } from "../../configs/routes";

const Projects = lazy(() => import("../../components/Projects/Projects"));
const SignForm = lazy(() => import("../../components/SignForm/SignForm"));

const LandingPage = memo((): JSX.Element => {
  const { getAll } = useProjects();
  const navigate = useNavigate();

  useEffect(() => {
    (async () => {
      await getAll(0);
    })();
  }, [getAll]);

  const projects = useAppSelector(({ projects }) =>
    projects.filter((project, index) => index < 6)
  );
  const isLogged = useAppSelector(({ user }) => user.isLogged);

  return (
    <>
      <HeroSectionStyled className="full-width">
        <div className="container">
          <h1 className="page__title page__title--landing">
            Let them know
            <span className="page__title--bold"> what you've got</span>
          </h1>

          <p className="page__title-subheading page__title-subheading--landing">
            Set your projects to a new level by sharing them with wideScope's
            glowing community of developers.
          </p>

          <Button
            renderAs="a"
            action={() => {
              navigate(navRoutes.signUp.path);
            }}
            children="Get started"
          />

          <Button
            customStyle="outline"
            renderAs="a"
            action={() => {
              navigate(navRoutes.explore.path);
            }}
            children="Explore"
          />
        </div>
      </HeroSectionStyled>

      <Suspense fallback={<p>Getting some projects...</p>}>
        <section className="langing-page__projects">
          <Projects projects={projects}></Projects>
        </section>
      </Suspense>

      <Suspense fallback={<p>Loading...</p>}>
        <CTASectionStyled className="full-width">
          <div className="container">
            {!isLogged && (
              <div className="landing-page__log-in">
                <div className="cta-section">
                  <h2 className="page__title">
                    Start sharing your projects. <br />
                    <span className="page__title--bold"> It's free</span>
                  </h2>
                </div>

                <SignForm isLogin={true} />
              </div>
            )}

            {isLogged && (
              <div className="landing-page__create">
                <div className="cta-section">
                  <h2 className="page__title">
                    Share your projects. <br />
                    <span className="page__title--bold"> It's free</span>
                  </h2>

                  <Button
                    customStyle="outline-invert"
                    renderAs="a"
                    action={() => {
                      navigate(navRoutes.createProject.path);
                    }}
                  >
                    Post a project
                  </Button>
                </div>
              </div>
            )}
          </div>
        </CTASectionStyled>
      </Suspense>
    </>
  );
});

export default LandingPage;
