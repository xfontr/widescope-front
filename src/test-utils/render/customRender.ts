import { render } from "@testing-library/react";
import { Wrapper } from "./Wrapper";

const customRender = (view: JSX.Element, ...options: any) =>
  render(view, { wrapper: Wrapper, ...options });

export * from "@testing-library/react";

export { customRender as render };
