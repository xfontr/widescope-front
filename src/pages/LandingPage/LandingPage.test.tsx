import { render as reactRender } from "@testing-library/react";
import { navRoutes } from "../../configs/routes";
import userEvent from "@testing-library/user-event";
import mockProject from "../../test-utils/mocks/mockProject";
import { render, screen } from "../../test-utils/render/customRender";
import { WrapperWithMockStore } from "../../test-utils/render/Wrapper";
import LandingPage from "./LandingPage";
import { act } from "react-dom/test-utils";

const mockNavigate = jest.fn().mockReturnThis();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockNavigate,
}));

describe("Given a LandingPage component", () => {
  describe("When intantiated if the user is not logged", () => {
    test("Then it should show a hero section, a list of projects and a CTA to log in", async () => {
      // eslint-disable-next-line testing-library/no-unnecessary-act
      await act(() => {
        render(<LandingPage />);
      });

      const landingPage = [
        screen.getByRole("heading", {
          name: "Let them know what you've got",
          level: 1,
        }),
        await screen.findByRole("heading", {
          name: "Start sharing your projects. It's free",
          level: 2,
        }),
        screen.getByText(
          "Set your projects to a new level by sharing them with wideScope's glowing community of developers."
        ),
        screen.getByText("Explore"),
        screen.getByText("Get started"),
      ];

      const projects = await screen.findAllByRole("heading", {
        name: mockProject.name,
        level: 3,
      });

      const loginForm = screen.getByText(
        "Share your latests projects with the WideScope community."
      );

      landingPage.forEach((element) => expect(element).toBeInTheDocument());
      projects.forEach((project) => expect(project).toBeInTheDocument());
      expect(loginForm).toBeInTheDocument();
    });
  });

  describe("When instantiated if the user is logged", () => {
    test("Then it should appear a create project CTA instead of the login form", () => {
      reactRender(<LandingPage />, { wrapper: WrapperWithMockStore });

      const ctaHeading = screen.getByRole("heading", {
        name: "Share your projects. It's free",
        level: 2,
      });
      const ctaButton = screen.getByText("Post a project");
      const loginForm = screen.queryByText(
        "Share your latests projects with the WideScope community."
      );

      expect(ctaHeading).toBeInTheDocument();
      expect(ctaButton).toBeInTheDocument();
      expect(loginForm).not.toBeInTheDocument();
    });

    test("Then the 'Post a project' button should send the user to the create page on click", async () => {
      reactRender(<LandingPage />, { wrapper: WrapperWithMockStore });

      const ctaButton = screen.getByText("Post a project");

      await userEvent.click(ctaButton);

      expect(mockNavigate).toHaveBeenCalledWith(navRoutes.createProject.path);
    });
  });

  describe("When instantiated", () => {
    test("If the user clicks the explore link, it should redirect to said page", async () => {
      reactRender(<LandingPage />, { wrapper: WrapperWithMockStore });

      const ctaButton = screen.getByText("Explore");

      await userEvent.click(ctaButton);

      expect(mockNavigate).toHaveBeenCalledWith(navRoutes.explore.path);
    });

    test("If the user clicks the Get started link, it should redirect to the sign up page", async () => {
      reactRender(<LandingPage />, { wrapper: WrapperWithMockStore });

      const ctaButton = screen.getByText("Get started");

      await userEvent.click(ctaButton);

      expect(mockNavigate).toHaveBeenCalledWith(navRoutes.signUp.path);
    });
  });
});
