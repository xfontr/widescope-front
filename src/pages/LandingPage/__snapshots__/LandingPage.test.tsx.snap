// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a LandingPage component When intantiated if the user is not logged Then it should match the snapshot 1`] = `
Array [
  <section
    className="sc-gsnTZi cPBToR full-width"
  >
    <div
      className="container"
    >
      <h1
        className="page__title page__title--landing"
      >
        Let them know
        <span
          className="page__title--bold"
        >
           what you've got
        </span>
      </h1>
      <p
        className="page__title-subheading page__title-subheading--landing"
      >
        Set your projects to a new level by sharing them with wideScope's glowing community of developers.
      </p>
      <a
        className="sc-bczRLJ jZmJeh undefined button--default "
        onClick={[Function]}
        rel="noopener noreferrer"
        target="_blank"
      >
        Get started
      </a>
      <a
        className="sc-bczRLJ jZmJeh undefined button--outline "
        onClick={[Function]}
        rel="noopener noreferrer"
        target="_blank"
      >
        Explore
      </a>
    </div>
  </section>,
  <p>
    Getting some projects...
  </p>,
  <p>
    Loading...
  </p>,
]
`;
